README.md
# LXDK8S - Kubernetes on LXD 
LXDK8S is a work inspired and adapted after the labs `Kubernetes The Hard Way` of [Kelsey Hightower](https://github.com/kelseyhightower/kubernetes-the-hard-way) .

As the original labs, the purpose of this work is to provide a playground to learn Kubernetes. Instead to use the Google Cloud Platform, LXDK8S runs on top of the UBUNTU/LXD infrastucture of your local machine.

All the orginal **Kubernetes The Hard Way** labs have been adapted to fit with the LXC commands and the limitations of a local machine. For speed up the installation process, the labs have been fully automatized using BASH scripts. Indeed, the organization of these scripts follows the sequence of the orginal labs of **Kelsey Hightower**, so please refer to his [documenation](https://github.com/kelseyhightower/kubernetes-the-hard-way/tree/master/docs) for understanding.

LXDK8S is updated to deliver a Kubernetes v1.14.2 playground.

## Cluster Details
```
                                 lxdk8s-lb
                           +------------------+
                           |  Load Balancer   | (HAProxy)
                           +------------------+

      lxdk8s-m0                  lxdk8s-m1                  lxdk8s-m2
+------------------+       +------------------+       +------------------+
|     Master-0     |       |     Master-1     |       |     Master-2     | (API-Server)
+------------------+       +------------------+       +------------------+ (Control-Manager)
|       Etcd       |       |       Etcd       |       |       Etcd       | (Scheduler)
+------------------+       +------------------+       +------------------+


      lxdk8s-w0                  lxdk8s-w1                  lxdk8s-w2
+------------------+       +------------------+       +------------------+
|     Worker-0     |       |     Worker-1     |       |     Worker-2     | (Kube-Proxy)
+------------------+       +------------------+       +------------------+ (KubeLet)
|  10.200.0.0/24   |       |  10.200.1.0/24   |       |  10.200.2.0/24   | (CNI) (CRI)
+------------------+       +------------------+       +------------------+ (Containerd)
|       Pods       |       |       Pods       |       |       Pods       | (RunC) (gVisor)
+------------------+       +------------------+       +------------------+ 
```
## Prerequisites
LXDK8S has been successfully installed on a local machine running UBUNTU 18.04 and LXD 3.13

You need a machine with at least 16 GB of RAM, and 8 CPUs.

Other combination of LINUX distributions and of LXD versions can probably work but it is not garantited

> Note: The LXD images used for the machines of the LXDK8S cluster are based on the UBUNTU 18.04 image.

#### 1. Install LXD 
On UBUNTU 18.04 machine installs LXD using the APT method
```
$ sudo apt update && apt upgrade
$ sudo apt install lxd
```

or, using the SNAP method.
```
$ sudo snap install lxd
```

Verify the LXD version is 3.13 or higher.
```
$ lxc version
```

#### 2. Init LXD
Before to start, you have to initialize the LXD infrastucture. 

To complete the LXD initialisation, the following command requires to set several parameters.
You can safely accept the default values.
```
$ sudo lxd init
```
#### 3. Proxy case
If your local machine is behind a proxy, make sure to set the `http_proxy` environment varibale properly before to start the installation process.

```
export http_proxy=http://<login>:<password>@<your.proxy.url>:3128/
```

LXDK8S takes care of this variable to set accordingly the proxy for LXD, APT, and CONTAINERD respectively. (see the `./config/lxdk8s.proxy` file)

## Installation
#### 1. Clone the repositorty
```
$ git clone https://github.com/i3oc9i/lxdk8s.git

$ cd ./lxdk8s
```

#### 2. Bootstrap the cluster
To create the LXDK8S cluster you can execute each task separately, or  
```
$ ./00-bootstrap-lxd.sh
$ ./01-get-client-tools.sh
$ ./10-generate-certs.sh
$ ./11-generate-kubeconfigs.sh
$ ./12-generate-encryption-keys.sh
$ ./20-bootstrap-haproxy.sh
$ ./30-bootstrap-etcd.sh
$ ./31-bootstrap-control-plane.sh
$ ./32-configure-rbac-kubelet-authorization.sh
$ ./40-bootstrap-workers.sh
$ ./50-create-network-routes.sh
```
use the `lxdk8s.bootstrap.sh` command. 
```
$ ./lxdk8s.bootstrap.sh
```
#### 3. Create the network routes
Use the `lxdk8s.net` command to create the network routes to let the pods to comunicate across the workers.
```
$ sudo ./bin/lxdk8s.net
```

#### 4. Save the lxdk8s.kubeconfig
During the bootstrap process the `lxdk8s.kubeconfig` configuration file is generated in the `./setup` directory. 

You can copy it in your `.kube` directory, or
```
$ mkdir -p ~/.kube
$ cp ./setup/lxdk8s.kubeconfig ~/.kube/config
```
set the `KUBECONFIG` environment variable into your `~/.bashrc` file.
```
export KUBECONFIG="~/lxdk8s/setup/lxdk8s.kubeconfig"
```

#### 5. Verify the cluster
Verify the components of the cluster.
```
$ ./bin/lxdk8s.kubectl --kubeconfig ./setup/lxdk8s.kubeconfig  get componentstatuses

NAME                 STATUS    MESSAGE             ERROR
scheduler            Healthy   ok                  
controller-manager   Healthy   ok                  
etcd-2               Healthy   {"health":"true"}   
etcd-0               Healthy   {"health":"true"}   
etcd-1               Healthy   {"health":"true"}
```

Verify the status of the nodes.
```
$ ./bin/lxdk8s.kubectl --kubeconfig ./setup/lxdk8s.kubeconfig  get node -o wide

NAME        STATUS   ROLES    AGE   VERSION   INTERNAL-IP      EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
lxdk8s-w0   Ready    <none>   77m   v1.14.2   10.234.143.181   <none>        Ubuntu 18.04.2 LTS   4.15.0-50-generic   containerd://1.2.6
lxdk8s-w1   Ready    <none>   76m   v1.14.2   10.234.143.26    <none>        Ubuntu 18.04.2 LTS   4.15.0-50-generic   containerd://1.2.6
lxdk8s-w2   Ready    <none>   75m   v1.14.2   10.234.143.86    <none>        Ubuntu 18.04.2 LTS   4.15.0-50-generic   containerd://1.2.6
```

Retrive the API URL address of the cluster
```
$ ./bin/lxdk8s.kubectl --kubeconfig ./setup/lxdk8s.kubeconfig  cluster-info  

Kubernetes master is running at https://10.234.143.198:6443

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
```

## Install Addons

#### CoreDNS
Use the following command to deploy CoreDNS.
```
$ ./addon/deploy-coredns.sh
```

## Using LXDK8S
After the bootstrap process is terminated the `./bin` directory will contains the follwing commands to control the cluster.  

```
lxdk8s.start
lxdk8s.status
lxdk8s.stop
lxdk8s.net
lxdk8s.kubectl
```
Add this directory at your `PATH`, or copy its content into the `/usr/local/bin`, or `~/.local/bin` directory.

#### 1. lxdk8s.start
To start the LXDK8S cluster.
```
$ ./bin/lxdk8s.start
```
#### 2. lxdk8s.stop
To stop the LXDK8S cluster.
```
$ ./bin/lxdk8s.stop
```
#### 3. lxdk8s.status
To retrive the status of the LXDK8S cluster.
```
$ ./bin/lxdk8s.status
>>> LXDK8S machines

| lxdk8s-lb | RUNNING | 10.234.143.198 (eth0) |
| lxdk8s-m0 | RUNNING | 10.234.143.49 (eth0)  |
| lxdk8s-m1 | RUNNING | 10.234.143.190 (eth0) |
| lxdk8s-m2 | RUNNING | 10.234.143.135 (eth0) |
| lxdk8s-w0 | RUNNING | 10.234.143.181 (eth0) |
| lxdk8s-w1 | RUNNING | 10.234.143.26 (eth0)  |
| lxdk8s-w2 | RUNNING | 10.234.143.86 (eth0)  |

>>> Load balancer status
tcp        0      0 10.234.143.198:6443     0.0.0.0:*               LISTEN      263/haproxy         
tcp        0      0 10.234.143.198:1936     0.0.0.0:*               LISTEN      263/haproxy         
>>> Kubernetes Control Plane status
NAME                 STATUS    MESSAGE             ERROR
scheduler            Healthy   ok                  
controller-manager   Healthy   ok                  
etcd-2               Healthy   {"health":"true"}   
etcd-1               Healthy   {"health":"true"}   
etcd-0               Healthy   {"health":"true"}   
>>> Kubernetes Nodes status
NAME        STATUS   ROLES    AGE   VERSION
lxdk8s-w0   Ready    <none>   88m   v1.14.2
lxdk8s-w1   Ready    <none>   87m   v1.14.2
lxdk8s-w2   Ready    <none>   86m   v1.14.2
```

#### 4. lxdk8s.net
To creates the network routes to let the pod to comunicate across the workers.
```
$ sudo ./bin/lxdk8s.net
$ ip route

default via 192.168.1.1 dev enp0s31f6 proto dhcp metric 100 
10.200.0.0/24 via 10.234.143.181 dev lxdbr0 
10.200.1.0/24 via 10.234.143.26 dev lxdbr0 
10.200.2.0/24 via 10.234.143.86 dev lxdbr0 
10.234.143.0/24 dev lxdbr0 proto kernel scope link src 10.234.143.1
... 
```

#### 6. lxdk8s.kubectl
To control the cluster.
```
$ ./bin/lxdk8s.kubectl --kubeconfig ./setup/lxdk8s.kubeconfig get all --all-namespaces

NAMESPACE     NAME                          READY   STATUS    RESTARTS   AGE
kube-system   pod/coredns-b4c55b465-nsdfm   1/1     Running   0          13m
kube-system   pod/coredns-b4c55b465-q8m2m   1/1     Running   0          13m

NAMESPACE     NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
default       service/kubernetes   ClusterIP   10.32.0.1    <none>        443/TCP                  39m
kube-system   service/kube-dns     ClusterIP   10.32.0.10   <none>        53/UDP,53/TCP,9153/TCP   13m

NAMESPACE     NAME                      READY   UP-TO-DATE   AVAILABLE   AGE
kube-system   deployment.apps/coredns   2/2     2            2           13m

NAMESPACE     NAME                                DESIRED   CURRENT   READY   AGE
kube-system   replicaset.apps/coredns-b4c55b465   2         2         2       13m
```

## Clean UP
To delete all the LXDK8S machines and remove all file generated in the `./setup` directory.
```
$ ./99-cleanup.sh
```


CHANGELOG
v1.0.0 - Kubernetes 1.14.2 and add CoreDNS 1.5.0 support

v1.0.1 - Fix proxy setting for LXD, APT, and CONTAINERD

v1.0.2 - Fix CONTAINERD_PROXY setting when value is 'noproxy'


40-bootstrap-workers.sh
#!/bin/bash
# Copyright (c) 2019 Ivano Coltellacci. All rights reserved.
# Use of this source code is governed by the MIT License that can be found in
# the LICENSE file.
set -eu

source ./config/lxdk8s.config

mkdir -p ./setup; cd setup

echo "(-) Download the Kubernetes Worker Node binaries"
# hack the file name schema change in the CNI versions
wget -q --show-progress --https-only --continue \
  https://github.com/containerd/containerd/releases/download/v${CONTAINERD_VER}/containerd-${CONTAINERD_VER}.linux-amd64.tar.gz \
  https://github.com/containernetworking/plugins/releases/download/${CNI_VER}/cni-plugins-linux-amd64-${CNI_VER}.tgz \
  https://github.com/kubernetes-sigs/cri-tools/releases/download/${CRI_VER}/crictl-${CRI_VER}-linux-amd64.tar.gz \
  https://github.com/opencontainers/runc/releases/download/${RUNC_VER}/runc.amd64 \
  https://storage.googleapis.com/gvisor/releases/nightly/${GVISOR_VER}/runsc \
  https://storage.googleapis.com/kubernetes-release/release/${K8S_VER}/bin/linux/amd64/kube-proxy \
  https://storage.googleapis.com/kubernetes-release/release/${K8S_VER}/bin/linux/amd64/kubelet \
  https://storage.googleapis.com/kubernetes-release/release/${K8S_VER}/bin/linux/amd64/kubectl 

chmod +x runc.amd64 runsc kube-proxy kubelet kubectl 
rm -f runc; ln -s runc.amd64 runc

for instance in lxdk8s-w0 lxdk8s-w1 lxdk8s-w2; do

echo ">>> Bootstraping the Kubernetes ${instance}"
echo "(0) Upgrade and install the OS dependencies of ${instance}"
if [ ${APT_PROXY} != "noproxy" ] ; then
cat > ubuntu-apt-proxy.conf <<EOF
Acquire::http::Proxy  "${APT_PROXY}";
Acquire::https::Proxy "${APT_PROXY}";
EOF

lxc file push ubuntu-apt-proxy.conf ${instance}/etc/apt/apt.conf.d/proxy.conf
fi

lxc exec ${instance} -- apt -y update
lxc exec ${instance} -- apt -y upgrade
lxc exec ${instance} -- apt -y install socat conntrack ipset
lxc exec ${instance} -- apt -y autoremove

echo "(1) Install the node binaries of ${instance}"
lxc exec ${instance} --  mkdir -p /etc/cni/net.d \
                                  /opt/cni/bin \
                                  /etc/containerd \
                                  /var/lib/kubelet \
                                  /var/lib/kube-proxy \
                                  /var/lib/kubernetes \
                                  /var/run/kubernetes

lxc file push  runc runsc kube-proxy kubelet kubectl ${instance}/usr/local/bin/

cat containerd-${CONTAINERD_VER}.linux-amd64.tar.gz | lxc exec ${instance} -- tar xzf - -C /
cat cni-plugins-linux-amd64-${CNI_VER}.tgz  | lxc exec ${instance} -- tar xzf - -C /opt/cni/bin/
cat crictl-${CRI_VER}-linux-amd64.tar.gz | lxc exec ${instance} -- tar xzf - -C /usr/local/bin/

echo "(2) Configure CNI Networking of ${instance}"
POD_CIDR=${POD_CIDR_NET[${instance}]}

cat > ${instance}-bridge.conf <<EOF
{
    "cniVersion": "0.3.1",
    "name": "bridge",
    "type": "bridge",
    "bridge": "cnio0",
    "isGateway": true,
    "ipMasq": true,
    "ipam": {
        "type": "host-local",
        "ranges": [
          [{"subnet": "${POD_CIDR}"}]
        ],
        "routes": [{"dst": "0.0.0.0/0"}]
    }
}
EOF

lxc file push ${instance}-bridge.conf ${instance}/etc/cni/net.d/10-bridge.conf

cat > ${instance}-loopback.conf <<EOF 
{
    "cniVersion": "0.3.1",
    "type": "loopback"
}
EOF

lxc file push ${instance}-loopback.conf ${instance}/etc/cni/net.d/99-loopback.conf

echo "(3) Configure Containerd of ${instance}"
cat > ${instance}-containerd-config.toml <<EOF
[plugins]
  [plugins.cri.containerd]
    snapshotter = "overlayfs"
    [plugins.cri.containerd.default_runtime]
      runtime_type = "io.containerd.runtime.v1.linux"
      runtime_engine = "/usr/local/bin/runc"
      runtime_root = ""
    [plugins.cri.containerd.untrusted_workload_runtime]
      runtime_type = "io.containerd.runtime.v1.linux"
      runtime_engine = "/usr/local/bin/runsc"
      runtime_root = "/run/containerd/runsc"
    [plugins.cri.containerd.gvisor]
      runtime_type = "io.containerd.runtime.v1.linux"
      runtime_engine = "/usr/local/bin/runsc"
      runtime_root = "/run/containerd/runsc"
EOF

lxc file push ${instance}-containerd-config.toml  ${instance}/etc/containerd/config.toml

if [ ${CONTAINERD_PROXY} != "noproxy" ]; then
Environment_HTTP_PROXY=$(echo Environment="HTTP_PROXY=${CONTAINERD_PROXY}")
else
Environment_HTTP_PROXY=""
fi

cat > ${instance}-containerd.service <<EOF
[Unit]
Description=containerd container runtime
Documentation=https://containerd.io
After=network.target

[Service]
#ExecStartPre=/sbin/modprobe overlay -- LXD does not support overlay
ExecStart=/bin/containerd
Restart=always
RestartSec=5
Delegate=yes
KillMode=process
OOMScoreAdjust=-999
LimitNOFILE=1048576
LimitNPROC=infinity
LimitCORE=infinity
${Environment_HTTP_PROXY}

[Install]
WantedBy=multi-user.target
EOF

lxc file push ${instance}-containerd.service ${instance}/etc/systemd/system/containerd.service

echo "(4) Configure Kubelet of ${instance}"
lxc file push ${instance}-key.pem ${instance}.pem ${instance}/var/lib/kubelet/
lxc file push ${instance}.kubeconfig ${instance}/var/lib/kubelet/kubeconfig
lxc file push ca.pem ${instance}/var/lib/kubernetes/

cat > ${instance}-kubelet-config.yaml <<EOF
kind: KubeletConfiguration
apiVersion: kubelet.config.k8s.io/v1beta1
authentication:
  anonymous:
    enabled: false
  webhook:
    enabled: true
  x509:
    clientCAFile: "/var/lib/kubernetes/ca.pem"
authorization:
  mode: Webhook
clusterDomain: "cluster.local"
clusterDNS:
  - "10.32.0.10"
podCIDR: "${POD_CIDR}"
resolvConf: "/run/systemd/resolve/resolv.conf"
runtimeRequestTimeout: "15m"
tlsCertFile: "/var/lib/kubelet/${instance}.pem"
tlsPrivateKeyFile: "/var/lib/kubelet/${instance}-key.pem"
failSwapOn: false
EOF

lxc file push ${instance}-kubelet-config.yaml ${instance}/var/lib/kubelet/kubelet-config.yaml

cat > ${instance}-kubelet.service <<EOF
[Unit]
Description=Kubernetes Kubelet
Documentation=https://github.com/kubernetes/kubernetes
After=containerd.service
Requires=containerd.service

#add option '--fail-swap-on=false' because LXD swap setting depend on host setting
[Service]
ExecStart=/usr/local/bin/kubelet \\
  --config=/var/lib/kubelet/kubelet-config.yaml \\
  --container-runtime=remote \\
  --container-runtime-endpoint=unix:///var/run/containerd/containerd.sock \\
  --image-pull-progress-deadline=2m \\
  --kubeconfig=/var/lib/kubelet/kubeconfig \\
  --network-plugin=cni \\
  --register-node=true \\
  --fail-swap-on=false \\
  --v=2
Restart=on-failure
RestartSec=5

[Install]
WantedBy=multi-user.target
EOF

lxc file push ${instance}-kubelet.service ${instance}/etc/systemd/system/kubelet.service

echo "(5) Configure the Kubernetes Proxy of ${instance}"
lxc file push kube-proxy.kubeconfig ${instance}/var/lib/kube-proxy/kubeconfig

cat > ${instance}-kube-proxy-config.yaml <<EOF
kind: KubeProxyConfiguration
apiVersion: kubeproxy.config.k8s.io/v1alpha1
clientConnection:
  kubeconfig: "/var/lib/kube-proxy/kubeconfig"
mode: "iptables"
clusterCIDR: "10.200.0.0/16"
EOF

lxc file push ${instance}-kube-proxy-config.yaml ${instance}/var/lib/kube-proxy/kube-proxy-config.yaml

cat > ${instance}-kube-proxy.service <<EOF
[Unit]
Description=Kubernetes Kube Proxy
Documentation=https://github.com/kubernetes/kubernetes

[Service]
ExecStart=/usr/local/bin/kube-proxy \\
  --config=/var/lib/kube-proxy/kube-proxy-config.yaml
Restart=on-failure
RestartSec=5

[Install]
WantedBy=multi-user.target
EOF

lxc file push ${instance}-kube-proxy.service ${instance}/etc/systemd/system/kube-proxy.service

echo "(6) Start the Worker Services of ${instance}"
lxc exec ${instance} -- systemctl daemon-reload
lxc exec ${instance} -- systemctl enable containerd kubelet kube-proxy
lxc exec ${instance} -- systemctl start containerd kubelet kube-proxy

echo "(7) Install the admin.kubeconfig of ${instance}"
lxc file push admin.kubeconfig ${instance}/root/

done

echo "(-) Verify the worker deployment"
sleep 15
lxc exec lxdk8s-m0 -- kubectl get nodes --kubeconfig admin.kubeconfig



